Pass custom_data through assemblers to kernel functions.

sclaus2 opened the pull request:
Add void* custom_data member to integral_data struct and pass it to assembly functions (matrix, vector, scalar, lifting). Custom_data defaults back to nullptr if no custom_data is set explicitly (with set_custom_data()). The required changes are minimal 

- Form.h: Add custom_data to integral_data, add accessors
- assemble_matrix_impl.h: Pass custom_data through all functions 
- assemble_vector_impl.h: Pass custom_data through all functions
- assemble_scalar_impl.h: Pass custom_data through all functions
- Python bindings: Expose set_custom_data() and custom_data()
- Add test_custom_data.py with comprehensive tests

This enables integration kernels to receive additional per-cell user data which is essential for example for runtime quadrature rules. With this extensions users are no longer required to completely rewrite all assembly routines for customised C-kernels. 

For an example library that relies on custom_data and its integration into the assembler see 

https://github.com/sclaus2/runintgen

sclaus2 committed:
Pass custom_data through assemblers to kernel functions.

jhale commented:
I say probably because it's a performance critical part of the code - might need a bit of thought.

sclaus2 committed:
Use std::optional<void*> for custom_data instead of raw pointer

Replace void* with nullptr pattern with std::optional<void*> for
custom_data in the Form class and assembly functions. This is the
canonical modern C++ approach for representing optional values.

Changes:
- Form.h: integral_data::custom_data now uses std::optional<void*>
- Form.h: custom_data() returns std::optional<void*>
- Form.h: set_custom_data() accepts std::optional<void*>
- assemble_*_impl.h: Function parameters use std::optional<void*>
- assemble_*_impl.h: Kernel calls use .value_or(nullptr)
- Python bindings: Handle std::optional with None support
- Test: Update to expect None instead of 0 for unset custom_data

Benefits:
- Clearer intent: "optional value" vs "magic nullptr"
- Type safety: Distinguishes "no value" from "null pointer value"
- Pythonic: Returns None instead of 0 when not set
- Zero-cost: .value_or(nullptr) compiles to same code

jhale commented on python/test/unit/fem/test_custom_data.py:
```diff
@@ -0,0 +1,287 @@
+        coordinate_dofs = numba.carray(coords_, (3, 3), dtype=xdtype)
+
+        # Cast void* to float64* and read the scale value
+        typed_ptr = voidptr_to_float64_ptr(custom_data)
```
Do we have any more complex examples? For example, an array of float64s index by the current cell. This is what @chrisrichardson mentioned before "a kernel might need to know which cell it's running on".

jhale commented on python/dolfinx/wrappers/fem.cpp:
```diff
@@ -761,6 +761,31 @@ void declare_form(nb::module_& m, std::string type)
+          "custom_data",
+          [](const dolfinx::fem::Form<T, U>& self,
+             dolfinx::fem::IntegralType type, int id,
+             int kernel_idx) -> std::optional<std::uintptr_t>
```
Why are we returning `std::uintptr_t` and not `void*`? If there is a good reason there needs to be a comment.

jhale reviewed (changes requested):
Looks pretty good - a few small changes to make.

sclaus2 commented on python/test/unit/fem/test_custom_data.py:
```diff
@@ -0,0 +1,287 @@
+
+    # Verify the assembly used our custom values
+    # The offset should contribute to each DOF
+    assert la.norm(b) > 0
```
I changed the tests and added more comprehensive ones.

sclaus2 committed:
Pass custom_data at Form creation time (immutable Form pattern)

Address review comments to maintain Form immutability:
- Remove set_custom_data method from Form class
- Pass custom_data as 5th element of integrals tuple at Form creation
- Integrals tuple format: (id, kernel_ptr, entities, coeffs, custom_data)
- custom_data can be None (maps to std::nullopt) or a pointer (uintptr_t)

Pass cell index through entity_local_index for cell integrals:
- Cell integrals now receive &cell instead of nullptr for entity_local_index
- Enables per-cell data lookup in custom kernels via custom_data
- FFCx-generated kernels don't read entity_local_index for cells (return 0)

Safety documentation:
- Add warning about void pointer safety to Form.h and Python bindings
- Document that users must keep data alive while Form is in use

Update tests:
- Update test_custom_jit_kernels.py to use 5-element tuples
- Expand test_custom_data.py with per-cell material property example

sclaus2 committed:
add None as fifth argument to integrals

sclaus2 committed:
add runtime quadrature rule test

sclaus2 committed:
Make sure that just custom_data is passed. No other changes

sclaus2 committed:
fix ruff and add ffcx branch for custom_data ptr casting

sclaus2 committed:
fix clang-format

sclaus2 commented on python/test/unit/fem/test_custom_data.py:
```diff
@@ -0,0 +1,287 @@
+
+
+# Helper intrinsic to cast void* to a typed pointer for custom_data
+@numba.extending.intrinsic
```
Moved to ffcx see PR [https://github.com/FEniCS/ffcx/pull/804](url)

sclaus2 committed:
fix docs

sclaus2 committed:
fix clang

sclaus2 committed:
more doc fixes

sclaus2 committed:
Fix tests in parallel

sclaus2 committed:
Merge branch 'main' into sclaus2/pass_custom_data_to_assembler

jhale commented on python/test/unit/fem/test_custom_data.py:
```diff
@@ -0,0 +1,415 @@
+    return tabulate
+
+
+@pytest.mark.parametrize("dtype", [np.float64])
```
Would be good to test both np.float64 and np.float32.

sclaus2 replied:
done

jhale commented on python/test/unit/fem/test_custom_data.py:
```diff
@@ -0,0 +1,415 @@
+    assert L._cpp_object.custom_data(IntegralType.cell, 0, 0) is None
+
+
+@pytest.mark.parametrize("dtype", [np.float64])
```
Extend all tests to use float64 and float32.

sclaus2 replied:
done

jhale committed:
Merge branch 'main' into sclaus2/pass_custom_data_to_assembler

sclaus2 committed:
Merge remote-tracking branch 'origin/main' into sclaus2/pass_custom_data_to_assembler

# Conflicts:
#	cpp/dolfinx/fem/assemble_matrix_impl.h
#	cpp/dolfinx/fem/assemble_vector_impl.h
#	python/dolfinx/wrappers/fem.cpp

sclaus2 committed:
fix ruff check

sclaus2 committed:
Format custom data expression helper

sclaus2 committed:
clang formatting

sclaus2 committed:
fix import dolfinx in test_custom_data
